Hazelcast.Test/Hazelcast.Client.Test/ClientNearCacheTest.cs

```diff
@@ -14,132 +14,339 @@
+
+            var idleConfig = new NearCacheConfig()
+                .SetMaxIdleSeconds(1);
+            config.AddNearCacheConfig("nearCacheIdle*", idleConfig);
         }
 
-        internal static IMap<object, object> map;
+        private IMap<object, object> _map;
 
-        //
-        [SetUp]
-        public void Init()
+        private ClientNearCache GetNearCache<TK, TV>(IMap<TK, TV> map)
         {
-            map = Client.GetMap<object, object>("nearCachedMap-" + TestSupport.RandomString());
+            return (map as ClientMapProxy<TK, TV>).NearCache;
         }
 
-        [TearDown]
-        public static void Destroy()
+        private IData ToKeyData(object k)
         {
-            map.Clear();
-            map.Destroy();
+            return ((HazelcastClientProxy) Client).GetClient().GetSerializationService().ToData(k);
         }
 
         [Test]
-        public void TestNearCache() 
+        public void TestNearCache()
         {
-            for (int i = 0; i < 100; i++) 
+            for (var i = 0; i < 100; i++)
             {
-                map.Put("key" + i, "value" + i);
+                _map.Put("key" + i, "value" + i);
             }
-            long begin = Clock.CurrentTimeMillis();
-            for (int i = 0; i < 100; i++) {
-                map.Get("key" + i);
+            var begin = Clock.CurrentTimeMillis();
+            for (var i = 0; i < 100; i++)
+            {
+                _map.Get("key" + i);
             }
-            long firstRead = Clock.CurrentTimeMillis() - begin;
+            var firstRead = Clock.CurrentTimeMillis() - begin;
             begin = Clock.CurrentTimeMillis();
-            for (int i = 0; i < 100; i++) {
-                map.Get("key" + i);
+            for (var i = 0; i < 100; i++)
+            {
+                _map.Get("key" + i);
             }
-            long secondRead = Clock.CurrentTimeMillis() - begin;
+            var secondRead = Clock.CurrentTimeMillis() - begin;
             Assert.IsTrue(secondRead < firstRead);
         }
 
         [Test]
-        public void testNearCacheGetAll() 
+        public void TestNearCacheGet()
+        {
+            _map.Put("key", "value");
+            _map.Get("key");
+
+            var clientNearCache = GetNearCache(_map);
+
+            Assert.AreEqual("value", clientNearCache.Get(ToKeyData("key")));
+        }
+
+        [Test]
+        public void TestNearCacheGetAll()
         {
-            var keys=new List<object>();
-            for (int i = 0; i < 100; i++) 
+            var keys = new List<object>();
+            for (var i = 0; i < 100; i++)
             {
-                map.Put("key" + i, "value" + i);
+                _map.Put("key" + i, "value" + i);
                 keys.Add("key" + i);
             }
 
-            map.GetAll(keys);
-            var mapImpl = map as ClientMapProxy<object, object>;
-            var clientNearCache = mapImpl.NearCache;
+            _map.GetAll(keys);
```

Comment:
I remember we had problems with similar tests. 
The scenario that is causing tests were as follows.
- This getAll populates the near cache.
- An invalidation event caused by map.put comes and invalidates a key. 
- And this results in less than expected number of items in the cache. 

One solution is to open the client after put is done and its events are already fired. 
An example test. 
https://github.com/hazelcast/hazelcast/blob/master/hazelcast-client/src/test/java/com/hazelcast/client/map/ClientMapNearCacheTest.java#L516